### ISSUE TITLE
--on-process-exit not getting called within a --on-event function

### ISSUE THREAD

**Issue by Baspar** [fish-shell/fish-shell#8766]
When creating a function with `--on-process-exit` or `--on-job-exit` from within a `--on-event` function, the "callback" function never gets called.

Example:
```fish
function fish_init --on-event fish_prompt
    sleep 3 &
    function fish_callback --on-process-exit (jobs --last --pid)
      commandline -a "Callback DONE"
    end
end
```

Executing manually the inner code seems to work fine

### Info:
**Fish Version**: 3.3.1
**Terminal**: Kitty 
**System**: Tested on MacOS & Linux

**Comment by ridiculousfish** [comment]
Yeah, jobs triggered from within event handlers do not themselves post process-exit events. I made this change in f7e2e7d26be2253f14551a023a339fca136593d2 and while I don't remember the details I think it was to avoid an infinite recursion for on-process-exit events. It should be relaxed for background jobs at least and maybe more; this is definitely something that can be imnproved.

**Comment by Baspar** [comment]
I opened a partial PR to cover the background jobs in #9096, but it might need some additional polish
